controllers/bmcjob_controller.go

```diff
@@ -18,38 +18,207 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"strconv"
 
-	"k8s.io/apimachinery/pkg/runtime"
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	bmcv1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
 )
 
 // BMCJobReconciler reconciles a BMCJob object
 type BMCJobReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	client           client.Client
+	bmcClientFactory BMCClientFactoryFunc
+	bmcTaskHandler   BMCTaskHandler
+	logger           logr.Logger
+}
+
+// NewBMCJobReconciler returns a new BMCJobReconciler
+func NewBMCJobReconciler(client client.Client, bmcClientFactory BMCClientFactoryFunc, bmcTaskHandler BMCTaskHandler, logger logr.Logger) *BMCJobReconciler {
+	return &BMCJobReconciler{
+		client:           client,
+		bmcClientFactory: bmcClientFactory,
+		bmcTaskHandler:   bmcTaskHandler,
+		logger:           logger,
+	}
 }
 
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=bmc.tinkerbell.org,resources=bmcjobs/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the BMCJob object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+// Reconcile runs a BMCJob.
+// Creates the individual BMCTasks on the cluster.
+// Runs each Task in a BMCTask and updates the BMCJob status.
 func (r *BMCJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
You might consider restructuring the call paths to model clearer levels of abstraction.

E.g.

```
func Reconcile() {
  // Initial context establishment
  job := retrieveJob()
  if !isNoopReconciliation(job) {
    // return
  }

  // Additional data retrieval activities
  host, secret := retrieveBaseboardManagementConnection(job) 
  client := newClient(host, secret)

  // Do a thing
  return reconcile(job, client)
}
```

This is in contrast to the way the code is modeled presently:

```
func Reconcile() {
  // Initial context establishment 
  job := retrieveJob()
  if !isNoopReconciliation(job) {
    // return
  }

  // Do a thing
  return reconcile(job)
}

func reconcile(job) {
  // Additional data retrieval activities
  host, secret := retrieveBaseboardManagementConnection(job) 
  client := newClient(host, secret)
  
  // Now actually do the thing we said we were doing.
  reconcileSomethingSpecific(job, client)
}
```

The former behaves as an orchestration function while the latter splinters the data retrieval and orchestration somewhat.